packages/ember-testing/lib/test.js

```diff
@@ -296,11 +296,18 @@ function helper(app, name) {
       // It's the first async helper in current context
       lastPromise = fn.apply(app, args);
     } else {
-      // wait for last helper's promise to resolve
-      // and then execute
+      // wait for last helper's promise to resolve and then
+      // execute. To be safe, we need to tell the adapter we're going
+      // asynchronous here, because fn may not be invoked before we
+      // return.
+      Test.adapter.asyncStart();
       run(function() {
         lastPromise = Test.resolve(lastPromise).then(function() {
-          return fn.apply(app, args);
+          try {
+            return fn.apply(app, args);
```

Comment:
We should throw/reject the promise and RSVP's [`onErrorDefault`](https://github.com/emberjs/ember.js/blob/bd130ef77ffc2417a7504e638e1619dbd4f61507/packages/ember-runtime/lib/ext/rsvp.js#L66) would fail the test.  It might not be necessary to call `asyncEnd` after an error since the test will have failed (provided we throw/reject).